frame/nis/src/lib.rs

```diff
@@ -0,0 +1,937 @@
+}
+impl<T> FungibleMutate<T> for NoCounterpart<T> {
+	fn mint_into(_who: &T, _amount: u32) -> DispatchResult {
+		Ok(())
+	}
+	fn burn_from(_who: &T, _amount: u32) -> Result<u32, DispatchError> {
+		Ok(0)
+	}
+}
+impl<T> Convert<Perquintill, u32> for NoCounterpart<T> {
+	fn convert(_: Perquintill) -> u32 {
+		0
+	}
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::{FungibleInspect, FungibleMutate};
+	pub use crate::weights::WeightInfo;
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{
+			nonfungible::{Inspect as NonfungibleInspect, Transfer as NonfungibleTransfer},
+			Currency, Defensive, DefensiveSaturating,
+			ExistenceRequirement::AllowDeath,
+			OnUnbalanced, ReservableCurrency,
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+	use sp_arithmetic::{PerThing, Perquintill};
+	use sp_runtime::{
+		traits::{AccountIdConversion, Bounded, Convert, ConvertBack, Saturating, Zero},
+		TokenError,
+	};
+	use sp_std::prelude::*;
+
+	type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+	type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+		<T as frame_system::Config>::AccountId,
+	>>::PositiveImbalance;
+	type ReceiptRecordOf<T> = ReceiptRecord<
+		<T as frame_system::Config>::AccountId,
+		<T as frame_system::Config>::BlockNumber,
+	>;
+	type IssuanceInfoOf<T> = IssuanceInfo<BalanceOf<T>>;
+	type SummaryRecordOf<T> = SummaryRecord<<T as frame_system::Config>::BlockNumber>;
+	type BidOf<T> = Bid<BalanceOf<T>, <T as frame_system::Config>::AccountId>;
+	type QueueTotalsTypeOf<T> = BoundedVec<(u32, BalanceOf<T>), <T as Config>::QueueCount>;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Information on runtime weights.
+		type WeightInfo: WeightInfo;
+
+		/// Overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The treasury's pallet id, used for deriving its sovereign account ID.
+		#[pallet::constant]
+		type PalletId: Get<PalletId>;
+
+		/// Currency type that this works on.
+		type Currency: ReservableCurrency<Self::AccountId, Balance = Self::CurrencyBalance>;
+
+		/// Just the `Currency::Balance` type; we have this item to allow us to constrain it to
+		/// `From<u64>`.
+		type CurrencyBalance: sp_runtime::traits::AtLeast32BitUnsigned
+			+ codec::FullCodec
+			+ Copy
+			+ MaybeSerializeDeserialize
+			+ sp_std::fmt::Debug
+			+ Default
+			+ From<u64>
+			+ TypeInfo
+			+ MaxEncodedLen;
+
+		/// Origin required for auto-funding the deficit.
+		type FundOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The issuance to ignore. This is subtracted from the `Currency`'s `total_issuance` to get
+		/// the issuance with which we determine the thawed value of a given proportion.
+		type IgnoredIssuance: Get<BalanceOf<Self>>;
+
+		/// The accounting system for the fungible counterpart tokens.
+		type Counterpart: FungibleMutate<Self::AccountId>;
```

Comment:
We should think about if and how we are going to make these funds usable in governance. 

In pools, we faced a similar dilemma, and we chose to keep the accounting of "points" that you earn in return of joining a pool internal to the pallet. 

Our plan to make pool funds in the short term usable in governance is via `AccountVote::Split` from the pool bonded account (where the funds of all the members live).

But seeing the approach here, I am more and more inclined towards moving pools to also be backed by an instance of pallet-assets for accounting this. 

What we would need in the governance aspect is a way to express: other than the base `type Currency`, these other assets are also eligible for voting.

Reply:
They are not usable in governance.

Reply:
What's the rationale for it though? is it for technical reason? 

DOTs locked in NIS seem perfectly fine to be used in governance as far as I can see.

Reply:
It's as per the definition of the functionality and underlying economic relationship. Conceptually, the DOT goes out of existence for the period.